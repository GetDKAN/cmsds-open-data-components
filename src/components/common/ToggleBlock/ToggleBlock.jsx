import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import classNames from 'classnames'

export default function ToggleBlock ({
  customId = undefined,
  className = 'toggle-block',
  title,
  children,
  headingClasses = 'toggle-block-title',
  headingLevel = 2,
  innerClasses = 'toggle-block-inner',
  allowToggle = true,
  defaultClosed = false,
  toggleCallback = () => {},
  additionalBlockHeadingMarkup = <></>
}) {
  const [show, toggleShow] = useState(!defaultClosed)

  const ToggleHeading = `h${headingLevel}`

  let toggleBlockHeading = (
    <>
      <ToggleHeading className={headingClasses}>
        {title}
      </ToggleHeading>
      {additionalBlockHeadingMarkup}
    </>
  )

  if (allowToggle) {
    toggleBlockHeading = (
      <>
        <ToggleHeading className={headingClasses}>
          <button type='button' onClick={() => { toggleShow(!show); toggleCallback(!show) }} aria-expanded={show}>
            <FontAwesomeIcon
              icon={['fas', show ? 'chevron-down' : 'chevron-right']}
              size='1x'
              aria-hidden='true'
              role='presentation'
            />
            {title}
          </button>
        </ToggleHeading>
        {additionalBlockHeadingMarkup}
      </>
    )
  }

  return (
    <div
      id={customId}
      className={
        classNames('toggle-block-container', {
          [className]: className
        })
      }
    >
      {toggleBlockHeading}
      <div
        className={
          classNames('toggle-block-panel', {
            [innerClasses]: innerClasses,
            open: show,
            closed: !show
          })
        }
      >
        {children}
      </div>
    </div>
  )
}

ToggleBlock.propTypes = {
  /**
   * Custom id applied to wrapping element
   */
  customId: PropTypes.string,
  /**
   * Button/Header text
   */
  title: PropTypes.node.isRequired,
  /**
   * Panel content
   */
  children: PropTypes.node.isRequired,
  /**
   * Custom classes applied to header element
   */
  headingClasses: PropTypes.string,
  /**
   * Heading level applied to toggle button
   */
  headingLevel: PropTypes.number,
  /**
   * Custom classes applied to panel element
   */
  innerClasses: PropTypes.string,
  /**
   * 'true' makes the component an accordion element
   */
  allowToggle: PropTypes.bool,
  /**
   * Custom classes applied to wrapping element
   */
  className: PropTypes.string,
  /**
   * `true` sets the accordion element closed by default
   */
  defaultClosed: PropTypes.bool,
  /**
   * Callback function used when accordion button element is toggled.
   * Passes the `show` state value as an argument
   */
  toggleCallback: PropTypes.func,
  /**
   * Additional markup to be added after the header h2.
   * Used for data archive download links
   */
  additionalBlockHeadingMarkup: PropTypes.node
}
