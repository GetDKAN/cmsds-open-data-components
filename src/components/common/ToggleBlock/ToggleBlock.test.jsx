import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import ToggleBlock from './ToggleBlock'
import userEvent from '@testing-library/user-event'
import { waitFor } from '@testing-library/react'
import ReactDOMServer from 'react-dom/server'

const componentArgs = {
  customId: '',
  title: 'Donec eu libero sit amet quam egestas semper?',
  children: (
    <div>
      <p>Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean ultricies mi vitae est. Mauris placerat eleifend leo. Quisque sit amet est et sapien ullamcorper pharetra. Vestibulum erat wisi, condimentum sed, commodo vitae, ornare sit amet, wisi. Aenean fermentum, elit eget tincidunt condimentum, eros ipsum rutrum orci, sagittis tempus lacus enim ac dui. Donec non enim in turpis pulvinar facilisis. Ut felis. Praesent dapibus, neque id cursus faucibus, tortor neque egestas augue, eu vulputate magna eros eu erat. Aliquam erat volutpat. Nam dui mi, tincidunt quis, accumsan porttitor, facilisis luctus, metus</p>
    </div>
  ),
  headingClasses: 'toggle-block-title',
  innerClasses: 'toggle-block-inner',
  allowToggle: true,
  className: 'toggle-block',
  defaultClosed: false
}

const renderComponent = (args = componentArgs) => render(
  <ToggleBlock {...args} />
)

describe('ToggleBlock component.', () => {
  it('Matches snapshot.', () => {
    const renderedToggleBlock = renderer.create(
      <ToggleBlock {...componentArgs} />
    ).toJSON()

    expect(renderedToggleBlock).toMatchSnapshot()
  })

  it('Renders toggle header if \'allowToggle\' prop is \'true\'.', () => {
    const { container } = renderComponent()

    expect(container.querySelector('.toggle-block-title')).toBeInTheDocument()
    expect(container.querySelector('.toggle-block-title').textContent).toEqual(componentArgs.title)
  })

  it('Does not render toggle header if \'allowToggle\' prop is \'false\'.', () => {
    const { container } = renderComponent({
      ...componentArgs,
      allowToggle: false
    })

    expect(container.querySelector('.toggle-block-title > button')).toBeNull()
  })

  it('Renders toggle panel open if \'defaultClosed\' prop is \'false\'.', () => {
    const { container } = renderComponent()

    expect(container.querySelector('.toggle-block-inner')).toHaveClass('open')
    expect(container.querySelector('.toggle-block-inner').innerHTML).toEqual(ReactDOMServer.renderToStaticMarkup(componentArgs.children))
  })

  it('Renders toggle panel closed if \'defaultClosed\' prop is \'true\'.', () => {
    const { container } = renderComponent({
      ...componentArgs,
      defaultClosed: true
    })

    expect(container.querySelector('.toggle-block-inner')).toHaveClass('closed')
  })

  it('Toggle panel opens and closes when header button is clicked.', async () => {
    const toggleCallback = vi.fn()

    const { container } = renderComponent({
      ...componentArgs,
      toggleCallback: toggleCallback()
    })

    // Panel rendered open initially
    expect(container.querySelector('.toggle-block-title > button').getAttribute('aria-expanded')).toEqual('true')

    // Click to close panel
    await waitFor(() => userEvent.click(container.querySelector('.toggle-block-title > button')))

    // Make assertions
    expect(toggleCallback).toHaveBeenCalled()
    expect(container.querySelector('.toggle-block-inner')).toHaveClass('closed')
    expect(container.querySelector('.toggle-block-title > button').getAttribute('aria-expanded')).toEqual('false')

    // Click to open panel
    await waitFor(() => userEvent.click(container.querySelector('.toggle-block-title > button')))

    // Make assertions
    expect(toggleCallback).toHaveBeenCalled()
    expect(container.querySelector('.toggle-block-inner')).toHaveClass('open')
    expect(container.querySelector('.toggle-block-inner').innerHTML).toEqual(ReactDOMServer.renderToStaticMarkup(componentArgs.children))
    expect(container.querySelector('.toggle-block-title > button').getAttribute('aria-expanded')).toEqual('true')
  })
})
