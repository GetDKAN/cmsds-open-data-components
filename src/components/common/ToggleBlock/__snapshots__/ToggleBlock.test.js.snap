// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleBlock component. Matches snapshot. 1`] = `
<div
  className="toggle-block-container toggle-block"
  id=""
>
  <h2
    className="toggle-block-title"
  >
    <button
      aria-expanded={true}
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-chevron-down fa-1x "
        data-icon="chevron-down"
        data-prefix="fas"
        focusable="false"
        role="presentation"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      Donec eu libero sit amet quam egestas semper?
    </button>
  </h2>
  <div
    className="toggle-block-panel toggle-block-inner open"
  >
    <div>
      <p>
        Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean ultricies mi vitae est. Mauris placerat eleifend leo. Quisque sit amet est et sapien ullamcorper pharetra. Vestibulum erat wisi, condimentum sed, commodo vitae, ornare sit amet, wisi. Aenean fermentum, elit eget tincidunt condimentum, eros ipsum rutrum orci, sagittis tempus lacus enim ac dui. Donec non enim in turpis pulvinar facilisis. Ut felis. Praesent dapibus, neque id cursus faucibus, tortor neque egestas augue, eu vulputate magna eros eu erat. Aliquam erat volutpat. Nam dui mi, tincidunt quis, accumsan porttitor, facilisis luctus, metus
      </p>
    </div>
  </div>
</div>
`;
